File: ports/dotnet/src/UnicodeSecurity/ValidatedUtf8.cs
// Refinement type for bytes validated as strict RFC 3629 UTF-8.
//
// The validity claim is pinned at the module-boundary level: the only way to
// construct a ValidatedUtf8 is via the smart constructor Validate, which routes
// through the port's strict decoder state machine (Security.IsValidUtf8) —
// never through System.Text / Encoding.UTF8.
//
// Rationale: the ingestion layer is security-critical. A plain byte array
// field on a codec output type carries no claim about its UTF-8 validity —
// downstream consumers have to either re-validate or trust the producer.
// ValidatedUtf8 makes the claim module-level, so a downstream consumer that
// wants the raw bytes has to explicitly Unwrap — which reads as "I am consuming
// the RFC 3629 claim here". A transcription of ports/rust/src/validated_utf8.rs.

namespace UnicodeSecurity;

/// <summary>
/// A byte array that has been validated as strict RFC 3629 UTF-8. The
/// constructor is intentionally private; <see cref="Validate"/> is the only
/// blessed way to build a <see cref="ValidatedUtf8"/>.
/// </summary>
public sealed class ValidatedUtf8
{
    private readonly byte[] _bytes;

    private ValidatedUtf8(byte[] bytes) => _bytes = bytes;

    /// <summary>
    /// Validate a byte sequence and, on success, return a
    /// <see cref="ValidatedUtf8"/> carrying the RFC 3629 validity claim.
    /// Returns <c>null</c> when the bytes fail the strict state machine.
    /// </summary>
    public static ValidatedUtf8? Validate(IReadOnlyList<byte> bytes) =>
        Security.IsValidUtf8(bytes) ? new ValidatedUtf8(bytes.ToArray()) : null;

    /// <summary>Borrow the validated bytes.</summary>
    public IReadOnlyList<byte> AsBytes => _bytes;

    /// <summary>
    /// Consume the validity claim, returning the underlying bytes. After this
    /// call the validity claim is no longer carried at the module-boundary
    /// level — the caller owns the "these bytes are RFC 3629 valid" reasoning
    /// from here forward.
    /// </summary>
    public byte[] Unwrap() => (byte[])_bytes.Clone();
}

File: ports/ruby/lib/unicode_ruby/validated_utf8.rb
# frozen_string_literal: true

require_relative "utf8"

module UnicodeRuby
  # Refinement type for bytes validated as strict RFC 3629 UTF-8.
  #
  # The validity claim is pinned at the module-boundary level: the only way to
  # construct a ValidatedUtf8 is via the smart constructor {ValidatedUtf8.validate},
  # which routes through the strict decoder state machine. A downstream consumer
  # that wants the raw bytes has to explicitly {unwrap} — which reads as "I am
  # consuming the RFC 3629 claim here".
  class ValidatedUtf8
    # Conventionally private; use {ValidatedUtf8.validate}. +bytes+ is a byte
    # array (Array<Integer>).
    def initialize(bytes)
      @bytes = bytes
      freeze
    end

    # Validate +data+ and, on success, return a ValidatedUtf8 carrying the
    # RFC 3629 validity claim. Returns +nil+ when the bytes fail the strict
    # state machine.
    def self.validate(data)
      return nil unless Utf8.valid?(data)

      new(data)
    end

    # Borrow the validated bytes.
    def as_bytes
      @bytes
    end
  end

  module ValidatedUtf8Ops
    module_function

    # Consume the validity claim, returning the underlying bytes. After this
    # call the validity claim is no longer carried at the module-boundary
    # level — the caller owns the "these bytes are RFC 3629 valid" reasoning.
    def unwrap(validated)
      validated.as_bytes
    end
  end
end
